torch/testing/_internal/common_methods_invocations.py

```diff
@@ -2913,6 +2913,28 @@ def sample_inputs_diag(op_info, device, dtype, requires_grad, **kwargs):
 
     return samples + [vec_sample]
 
+def sample_inputs_diagonal(op_info, device, dtype, requires_grad, **kwargs):
+    make_arg = partial(make_tensor, dtype=dtype, device=device, requires_grad=requires_grad, low=None, high=None)
+
+    # 2D Tensors
+    tensors_2d = (
+        make_arg((M, M), low=None, high=None),
+        make_arg((3, 5), low=None, high=None),
+        make_arg((5, 3), low=None, high=None),
+    )
+
+    # 3D Tensors
+    tensors_3d = (
+        make_arg((M, M, M), low=None, high=None),
+    )
+
+    args_2d = ((), (2,), (-2,), (1,), (2,))
+    args_3d = ((1, 1, 2), (2, 0, 1), (-2, 0, 1))
+
+    tensors = [*product(tensors_2d, args_2d), *product(tensors_3d, args_3d)]
```

Comment:
I think using `product` with already materialised tensor is not good.

> I tried without the clone_inputs fuction in the patch. But it looks to be necessary as the sample_inputs for tile/repeat uses itertools.product

Ref: https://github.com/pytorch/pytorch/issues/52135#issuecomment-777523846

So I'd suggest taking product of `shape_2d` and `args_2d` and then iterating of the product and materializing unique tensor for each arg (like we do usually).

Not sure if it isn't an issue anymore.
**NOTE**: This will lead to more tensors being materialised than the current approach.
cc: @mruberry

Reply:
We should prioritize each sample input having an independent set of tensors that, if modified, won't affect other sample inputs. So I agree that taking the product over the non-tensor constructor args is preferable, even though it will generate more tensors. Reusing tensors in the test suite is a challenge I don't think we want (at least not now).

Reply:
This should be resolved in recent commit.